File: aspx/App_Code/Kpis.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SharePointPmDashboard.App_Code
{
    /// <summary>Pure helper functions computing dashboard KPIs, mirroring kpis.py.</summary>
    public static class Kpis
    {
        private static readonly HashSet<string> DoneStatuses = new HashSet<string>(
            new[] { "done", "completed", "closed" }, StringComparer.OrdinalIgnoreCase);

        public static double AveragePercentComplete(List<TaskItem> tasks)
        {
            if (tasks == null || tasks.Count == 0)
            {
                return 0.0;
            }
            return Math.Round(tasks.Average(t => t.PercentComplete) * 100, 1);
        }

        public static int OverdueCount(List<TaskItem> tasks, DateTime asOf)
        {
            if (tasks == null)
            {
                return 0;
            }
            return tasks.Count(t =>
                t.DueDate.HasValue &&
                t.DueDate.Value < asOf &&
                !DoneStatuses.Contains(t.Status ?? ""));
        }

        public static List<KeyValuePair<string, int>> TasksByStatus(List<TaskItem> tasks)
        {
            if (tasks == null)
            {
                return new List<KeyValuePair<string, int>>();
            }
            return tasks
                .GroupBy(t => t.Status ?? "")
                .Select(g => new KeyValuePair<string, int>(g.Key, g.Count()))
                .OrderByDescending(kv => kv.Value)
                .ToList();
        }

        public static List<KeyValuePair<string, int>> TasksByOwner(List<TaskItem> tasks)
        {
            if (tasks == null)
            {
                return new List<KeyValuePair<string, int>>();
            }
            return tasks
                .GroupBy(t => t.AssignedTo ?? "")
                .Select(g => new KeyValuePair<string, int>(g.Key, g.Count()))
                .OrderByDescending(kv => kv.Value)
                .ToList();
        }

        public static int OpenRiskCount(List<RiskItem> risks)
        {
            if (risks == null)
            {
                return 0;
            }
            return risks.Count(r => !DoneStatuses.Contains(r.Status ?? ""));
        }
    }
}

File: kpis.py
"""Pure functions computing dashboard KPIs from normalized task/risk DataFrames.

Kept separate from streamlit_app.py so they can be unit tested directly,
without going through the Streamlit UI layer.
"""

import pandas as pd

DONE_STATUSES = {"done", "completed", "closed"}


def average_percent_complete(tasks_df: pd.DataFrame) -> float:
    if tasks_df.empty or "percent_complete" not in tasks_df.columns:
        return 0.0
    return round(tasks_df["percent_complete"].mean() * 100, 1)


def overdue_count(tasks_df: pd.DataFrame, as_of: pd.Timestamp) -> int:
    if tasks_df.empty or "due_date" not in tasks_df.columns:
        return 0
    is_done = tasks_df["status"].str.lower().isin(DONE_STATUSES)
    is_overdue = tasks_df["due_date"].notna() & (tasks_df["due_date"] < as_of) & ~is_done
    return int(is_overdue.sum())


def tasks_by_status(tasks_df: pd.DataFrame) -> pd.DataFrame:
    if tasks_df.empty:
        return pd.DataFrame(columns=["status", "count"])
    counts = tasks_df["status"].value_counts().reset_index()
    counts.columns = ["status", "count"]
    return counts


def tasks_by_owner(tasks_df: pd.DataFrame) -> pd.DataFrame:
    if tasks_df.empty:
        return pd.DataFrame(columns=["assigned_to", "count"])
    counts = tasks_df["assigned_to"].value_counts().reset_index()
    counts.columns = ["assigned_to", "count"]
    return counts


def open_risk_count(risks_df: pd.DataFrame) -> int:
    if risks_df.empty or "status" not in risks_df.columns:
        return 0
    is_closed = risks_df["status"].str.lower().isin(DONE_STATUSES)
    return int((~is_closed).sum())
